## Pandas analysis

This exercise consists in analyzing a dataset containg timing information from a series of Time-to-Digital-Converters (TDC) implemented in a couple of FPGAs. Each measurement (i.e. each row of the input file) consists of a flag that specifies the type of message ('HEAD', which in this case is always 1), two addresses of the TDC providing the signal ('FPGA' and 'TDC_CHANNEL'), and the timing information ('ORBIT_CNT', 'BX_COUNTER', and 'TDC_MEAS'). Each TDC count corresponds to 25/30 ns, whereas a unit of BX_COUNTER corresponds to 25 ns, and the ORBIT_CNT is increased every 'x' BX_COUNTER. This allows to store the time in a similar way to hours, minutes and seconds.

In [1]:
# If you didn't download it yet, please get the relevant file now!
!wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ~/data/

'wget' is not recognized as an internal or external command,
operable program or batch file.


1\. Create a Pandas DataFrame reading N rows of the 'data_000637.txt' dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k.

In [38]:
import numpy as np
import pandas as pd

file_name="~/LaboratoryOfComputationalPhysics_Y5/data/data_000637.txt"

data=pd.read_csv(file_name)
print("Number of rows:", len(data.values))

N = 25000
print("Read", N, "rows")

df = data[:N]
df[:50]



Number of rows: 1310720
Read 25000 rows
         HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0           1     0          123  3869200167        2374        26
1           1     0          124  3869200167        2374        27
2           1     0           63  3869200167        2553        28
3           1     0           64  3869200167        2558        19
4           1     0           64  3869200167        2760        25
...       ...   ...          ...         ...         ...       ...
1310715     1     0           62  3869211171         762        14
1310716     1     1            4  3869211171         763        11
1310717     1     0           64  3869211171         764         0
1310718     1     0          139  3869211171         769         0
1310719     1     0           61  3869211171         762        18

[1310720 rows x 6 columns]


HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0      1     0          123  3869200167        2374        26
1      1     0          124  3869200167        2374        27
2      1     0           63  3869200167        2553        28
3      1     0           64  3869200167        2558        19
4      1     0           64  3869200167        2760        25
5      1     0           63  3869200167        2762         4
6      1     0           61  3869200167        2772        14
7      1     0          139  3869200167        2776         0
8      1     0           62  3869200167        2774        21
9      1     0           60  3869200167        2788         7
10     1     1            7  3869200167        2785         4
11     1     0           64  3869200167        2786        19
12     1     1            6  3869200167        2792        18
13     1     0           36  3869200167        2791        23
14     1     0           56  3869200167        2789         3
15     1     1          139  3869200167        2797         0
16     1     1            8  3869200167        2787        14
17     1     0           63  3869200167        2790        10
18     1     1            5  3869200167        2795         4
19     1     0           53  3869200167        2796        26
20     1     1           10  3869200167        2789        14
21     1     0           57  3869200167        2789        10
22     1     0           61  3869200167        2790        23
23     1     0           38  3869200167        2799        15
24     1     0           58  3869200167        2795        19
25     1     0           62  3869200167        2797        14
26     1     0           59  3869200167        2799        14
27     1     0           59  3869200167        3081        21
28     1     0           61  3869200167        3081         1
29     1     0           60  3869200167        3083        20
30     1     0          139  3869200167        3085         0
31     1     0           62  3869200167        3079         4
32     1     0           61  3869200167        3085        25
33     1     0           54  3869200167        3134         4
34     1     0           53  3869200167        3136        25
35     1     0           51  3869200167        3146        20
36     1     0           59  3869200167        3176         8
37     1     0           53  3869200167        3181         8
38     1     0           60  3869200167        3175        17
39     1     0          139  3869200167        3186         0
40     1     0           54  3869200167        3184        28
41     1     0           56  3869200167        3177        20
42     1     0           57  3869200167        3187        17
43     1     1            1  3869200168           7        18
44     1     0           64  3869200168           7        16
45     1     1            1  3869200168          13        18
46     1     0           63  3869200168           8        28
47     1     1            2  3869200168          17        26
48     1     0          139  3869200168          23         0
49     1     0           61  3869200168          18        17

2\. Find out the number of BX in a ORBIT (the value 'x').

In [31]:
x = df.max()[4]
print(x)

3563


3\. Find out how much the data taking lasted. You can either make an estimate based on the fraction of the measurements (rows) you read, or perform this check precisely by reading out the whole dataset.

In [55]:
# Assuming an approximately linear relationship between number of measurements taken
# and time passed, we can calculate total time from the fraction 
# 25'000/1310720 = len.data.values

i = df.iloc[0]
print(i)

f = df.iloc[-1]
print(f)

delta_orbit = f[3] - i[3]
delta_bxCount = f[4] - i[4]
delta_TDC = f[5] - i[5]

#t = np.zeros((3,))

#for j in range(0,3):
#    t[j] = f[j+3] - i[j+3]

# time passed in nanoseconds

t_fraction = delta_orbit*x*25 + delta_bxCount*25 + delta_TDC*25/30
#t_fraction1 = t[0]*x*25 + t[1]*25 + t[2]*25/30

print("Seconds passed: ", t_fraction/(10**9))

print("Total time elapsed (estimate based on fraction): ", t_fraction*len(data.values)/N/(10**9))

data.iloc

t_total = 

HEAD                    1
FPGA                    0
TDC_CHANNEL           123
ORBIT_CNT      3869200167
BX_COUNTER           2374
TDC_MEAS               26
Name: 0, dtype: int64
HEAD                    1
FPGA                    1
TDC_CHANNEL             3
ORBIT_CNT      3869200416
BX_COUNTER           2101
TDC_MEAS                6
Name: 24999, dtype: int64
Seconds passed:  0.022172833333333333
Total time elapsed (estimate based on fraction):  1.1624950442666666


4\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information).

5\. Replace the values (all 1) of the HEAD column randomly with 0 or 1.

6\. Create a new DataFrame that contains only the rows with HEAD=1.

7\. Make two occupancy plots (one for each FPGA), i.e. plot the number of counts per TDC channel

8\. Use the groupby method to find out the noisy channels, i.e. the TDC channels with most counts (say the top 3)

9\. Count the number of unique orbits. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139